csrc/val_graph_visitor.cpp

```diff
@@ -149,4 +149,387 @@ void ValGraphVisitor::traverse() {
+  // domains 1 and 2 are merged to produce domain 3, and then domains
+  // 0 and 3 are merged to produce domain 4.
+  //
+  // 0       1         2
+  //
+  // |       |         |
+  // |       |         |
+  // |       +-->   <--+
+  // |            3
+  // |            |
+  // |            |
+  // +----> 4 <---+
+  //
+  // Suppose we want to find the shortest path from {4} to {0, 1,
+  // 2}. The correct answer should be:
+  //
+  //   Backward merge of 0, 3 -> 4
+  //   Backward merge of 1, 2 -> 3
+  //
+  // However, the above traversal would produce a path of:
+  //
+  //   Backward merge of 0, 3 -> 4
+  //   Backward merge of 1, 2 -> 3
+  //   Backward merge of 1, 2 -> 3
+  //   Backward merge of 0, 3 -> 4
+  //
+  // This is because, since nodes 0, 1 and 2 are the starting nodes,
+  // we would first visit 4 from 0, and then 3 from 1 and again 3 from
+  // 2. Since node 3 would be visited twice, the path from 3 to 4
+  // would be traversed twice as well. Obviously, just reversing this
+  // path wouldn't give the correct path. There are two issues here:
+  //
+  // - The first visit to node 4 from node 0 should not be taken since
+  //   node 4 must appear after node 3
+  // - Visiting the same node multiple times is redundant and should
+  //   be removed
+  //
+  // Both problems could be solved by taking into considerations if
+  // nodes are ready to visit and also are already visited, just like
+  // done in the forward traversal. However, there's an additional
+  // complexity in this case because the following graph is also valid:
+  //
+  //         1         2
+  //
+  // |       |         |
+  // |       |         |
+  // |       +-->   <--+
+  // |            3
+  // |            |
+  // |            |
+  // +----> 4 <---+
+  //
+  // Notice that node 0 is missing, meaning the shortest path problem
+  // in this case is  from node 4 to nodes 1 and 2, and node 0 is not
+  // of interest. The correct path is still the same, i.e., first
+  // backward merge of 0 and 3 and then another backward merge of 1
+  // and 2. It is just node 0 is discarded as it is not of
+  // interest. In this case, however, if the
+  // traversal was enforced to honor the dependency of each node,
+  // it would not be able to visit the backward merge of 0 and 3 as
+  // node 0 is missing.
+  //
+  // A straightforward solution here is simply first generating the
+  // path as above and for each node, take the last visit only. Note
+  // that the last visit is always guaranteed to satisfy its
+  // dependencies.
+  //
+  // Here, instead of finding the last appearance of each node, the
+  // path is first reversed and then only the first appearance is
+  // taken since the path needs to be reversed anyway.
+  //
+  // See the ValGraphBFS2 test for a concrete example.
+
+  std::reverse(path.begin(), path.end());
+
+  VectorOfUniqueEntries<std::pair<ExprGroup, Direction>> unique_path(
+      path.begin(), path.end());
```

Comment:
Could you do this without moving elements in the path first?
```suggestion
  VectorOfUniqueEntries<std::pair<ExprGroup, Direction>> unique_path(
      path.rbegin(), path.rend());
```